File: 2021/day_06.js
/* --- Day 6: Lanternfish --- */

function numFish( t0, numDays){
    fish = new Array(9).fill(0);
    fish[t0] = 1;
    for( let d = 0; d < numDays; d++ ){
        fish[(d + 7) % fish.length] += fish[d % fish.length];
    }
    return sum(fish);
}

function day_06a( numDays = 80){
    let data = parseInput( Number, /,/),
        fish = Array.from( new Array(7), ( _, i) => numFish( i, numDays));
    return sum( data.map( x => fish[x]));
}

function day_06b(){
    return day_06a(256);
}

File: 2021/day_06.py
#!/usr/bin/python3

# read input as list integers seperated by ","
input = open( "day_06.input")
data = [ int(x) for x in input.readline().split(",")]
input.close()

# helper functions
def numFish( fish, numDays):
    for day in range(numDays):
        fish[(day + 7) % len(fish)] += fish[day % len(fish)];
    return sum(fish);

# solutions
def silver( numDays = 80):
    fish = [0] * 9
    for n in data:
        fish[n] += 1
    return numFish( fish, numDays)

def gold():
    return silver(256)
